src/kern/npf_alg_pptp.c

```diff
@@ -0,0 +1,788 @@
+ *
+ * Result:
+ *   NULL - no empty slots or slot to reuse
+ *   otherwise - a reference to a slot marked as used
+ *      and into which client_call_id and trans_client_call_id are written
+ */
+static pptp_gre_con_slot_t *
+npfa_pptp_gre_slot_lookup_and_use(npf_cache_t *npc, pptp_gre_conns_t *gre_conns,
+		  uint16_t client_call_id, uint16_t trans_client_call_id)
+{
+	pptp_gre_con_slot_t *slot, *empty_slot, old_reused_slot, new_slot;
+	bool reuse_slot;
+
+	reuse_slot = false;
+	empty_slot = NULL;
+
+	npfa_pptp_tcp_conn_lock(gre_conns);
+
+	/* scan all slots to ensure that there is no one using the call_id */
+	for (int i = 0; i < PPTP_MAX_GRE_PER_CLIENT; i++) {
+		slot = &gre_conns->gre_con_slots[i];
+		/* if call_id is already in use by a slot, then
+		 * expire associated GRE connection and reuse the slot
+		 */
+		if (slot->flags & PPTP_ALG_FL_FREE_SLOT) {
+			/* empty slot */
+			empty_slot = slot;
+		}
+		else if (slot->orig_client_call_id == client_call_id) {
+			reuse_slot = true;
+			old_reused_slot.u64 = slot->u64;
+			break;
+		}
+	}
+
+	/* use empty slot or reuse a slot with the same client_call_id */
+	if (reuse_slot || empty_slot != NULL) {
+		if (!reuse_slot)
+			slot = empty_slot;
+
+		new_slot.orig_client_call_id = client_call_id;
+		new_slot.ctx.client_call_id = trans_client_call_id;
+		new_slot.ctx.server_call_id = 0;
+		new_slot.flags = 0;
+
+		slot->u64 = new_slot.u64;
+	}
+
+	npfa_pptp_tcp_conn_unlock(gre_conns);
+
+	if (reuse_slot) {
+		npfa_pptp_gre_slot_free(npc->npc_ctx, &old_reused_slot,
+				  npc->npc_ips[NPF_SRC], npc->npc_ips[NPF_DST]);
+		return slot;
+	}
+
+	return empty_slot;
+}
+
+/*
+ * Lookup for an empty gre slot with original server call_id
+ */
+static pptp_gre_con_slot_t *
+npfa_pptp_gre_slot_lookup_with_server_call_id(pptp_gre_conns_t *gre_conns,
+		  uint16_t server_call_id)
+{
+	for (int i = 0; i < PPTP_MAX_GRE_PER_CLIENT; i++) {
+		pptp_gre_con_slot_t *gre_slot = &gre_conns->gre_con_slots[i];
+
+		if ((gre_slot->flags & PPTP_ALG_FL_FREE_SLOT) == 0 &&
+				  gre_slot->ctx.server_call_id == server_call_id)
+			return gre_slot;
+	}
+
+	return NULL;
+}
+
+/*
+ * Lookup for an empty gre slot with client call id
+ */
+static pptp_gre_con_slot_t *
+npfa_pptp_gre_slot_lookup_with_client_call_id(pptp_gre_conns_t *gre_conns,
+		  uint16_t call_id)
+{
+	for (int i = 0; i < PPTP_MAX_GRE_PER_CLIENT; i++) {
+		pptp_gre_con_slot_t *gre_slot = &gre_conns->gre_con_slots[i];
+
+		if ((gre_slot->flags & PPTP_ALG_FL_FREE_SLOT) == 0 &&
+				  gre_slot->ctx.client_call_id == call_id)
+			return gre_slot;
+	}
+
+	return NULL;
+}
+
+/*
+ * Init and setup new alg arg
+ */
+static pptp_gre_conns_t *
+npfa_pptp_gre_conns_alloc(npf_nat_t *nt)
+{
+	pptp_gre_conns_t *res_gre_conns;
+	pptp_gre_conns_t *new_gre_conns;
+
+	new_gre_conns = npfa_pptp_gre_conns_init();
+	if (new_gre_conns == NULL)
+		return NULL;
+
+	res_gre_conns = npf_nat_cas_alg_arg(nt, (uintptr_t)NULL,
```

Comment:
Err.. shouldn't we associate the ALG and set its argument at the same time? Generally, they should go together, but it seems to be tricky here. Having `npf_nat_cas_alg_arg()` is a a bit dodgy..

Reply:
If moved memory allocation of the gre slots to npfa_pptp_tcp_match() would it eliminate concurrency? npf_alg_match() is called by npf_do_nat() and I don't see any locks that would prevent a race there?

Reply:
Another argument is that there is a small chance that some rogue TCP connections will not get to the point where GRE slot's memory is needed, so I think it's good idea to allocate memory when it's really needed. What's wrong with npf_nat_cas_alg_arg? Why is it dodgy? I think it's pretty much straightforward :)